### PR TITLE
Refactor some core pieces of chef_objects.

### PR DESCRIPTION
**Pull request by sersut**

### CODE REVIEW

File: src/chef_object_core.erl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                            ej_array_map()       |
+                            ej_object_map()      |
+                            {[ej_json_val_spec()]}.
+
+%% Call this instead of ej:valid() if you want to validate specs
+%% and check for invalid top-level keys. This will not check for
+%% invalid keys beyond the top-level.
+-spec strictly_valid(Constraints :: ej_json_spec(), ValidKeys :: [binary()],  Ejson :: json_object()) -> ok | #ej_invalid{}.
+strictly_valid(Constraints, ValidKeys, Ejson) ->
+    case allowed_keys(ValidKeys, Ejson) of
```
Comment (jamesc) [inline src/chef_object_core.erl:58]: Why the case?  I would just do:

```
% allowed_keys will throw, not return
ok = allowed_keys(ValidKeys, Ejson),
ej:valid(Constraints, Ejson).
```

File: src/chef_object_core.erl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    end.
+
+allowed_keys(_ValidKeys, []) ->
+    ok;
+allowed_keys(ValidKeys, {List}) when is_list(List) ->
+    allowed_keys(ValidKeys, List);
+allowed_keys(ValidKeys, [{Item, _}|Rest]) ->
+    case lists:member(Item, ValidKeys) of
+        true -> allowed_keys(ValidKeys, Rest);
+        _ ->
```
Comment (jamesc) [inline src/chef_object_core.erl:71]: Why not just `false` here, since lists:member only returns true/false.

**Comment by jamesc** [comment]
Are there associated deletes as well, or is this duplicating code for now ?

File: src/chef_object_core.erl [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+%%
+
+%% @doc General utility module for common functions that operate on
+%% "Chef Objects", such as nodes, roles, etc.
+-module(chef_object_core).
+
+-include("chef_types.hrl").
+-include_lib("ej/include/ej.hrl").
+
+-export([strictly_valid/3
```
Comment (seth) [inline src/chef_object_core.erl:28]: Perhaps we should move the strictly_valid functionality into ej. Doesn't seem to be anything chef object specific here.

File: src/chef_regex_core.erl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+
+-module(chef_regex_core).
```
Comment (seth) [inline src/chef_regex_core.erl:25]: An aside: it's time for us to pull in the parse transform so that our regex's can be compiled once at compile time.

Comment (seth) [inline src/chef_regex_core.erl:25]: And we may actually want to revisit some of the various refactoring that has taken place as well as this in light of wanting to do that.

**Comment by marcparadise** [comment]
These would need to be recreated against current code in oc_erchef

**Pull request closed without merging**